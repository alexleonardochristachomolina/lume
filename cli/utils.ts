import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { isUrl } from "../core/utils/path.ts";
import { getConfigFile } from "../core/utils/lume_config.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(config?: string): Promise<Site> {
  let url: string | undefined;
  const watchInclude = [];

  if (config && isUrl(config)) {
    url = config;
  } else {
    const path = await getConfigFile(config);

    if (path) {
      watchInclude.push(path);
      url = toFileUrl(path).href;
    }
  }

  if (url) {
    log.info(`Loading config file <gray>${url}</gray>`);
    const mod = await import(url);
    const site = mod.default as Site | undefined;
    if (!(site instanceof Site)) {
      log.critical(
        `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
      );
      throw new Error("Site instance is not found");
    }

    site._data.configFile = url;
    site.options.watcher.include.push(...watchInclude);
    return site;
  }

  return lume();
}

/** Create a site intance and build it */
export async function buildSite(config?: string): Promise<Site> {
  const site = await createSite(config);

  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );

  return site;
}
