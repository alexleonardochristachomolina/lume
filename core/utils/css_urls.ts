type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    result += `${code.slice(index, start)}"${newUrl}"`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    const end = code.indexOf(")", start);
    let url = code.slice(start, end).trim();
    // Remove quotes
    if (url.startsWith('"') || url.startsWith("'")) {
      url = url.slice(1, -1);
    }
    return ["url", url, start, end];
  }
}
