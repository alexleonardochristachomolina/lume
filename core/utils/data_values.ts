import type { Data } from "../file.ts";

/**
 * Get the value of a page data
 * For example, if the value is "=title", it returns the value of the page data "title"
 * If the value is "$.title", it will return the value of the element with the selector ".title"
 */
export function getDataValue(data: Partial<Data>, value?: unknown) {
  // Get the value from the page data
  if (typeof value === "string") {
    return searchValue(data, value);
  }

  if (typeof value === "function") {
    return value(data);
  }

  return value;
}

function searchValue(data: Partial<Data>, value: string): unknown {
  if (!value) {
    return;
  }

  if (value.startsWith("=")) {
    let key = value.slice(1);
    [key, value] = parseFallback(key);

    if (!key.includes(".")) {
      return data[key] ?? searchValue(data, value);
    }

    const keys = key.split(".");
    let val = data;
    for (const key of keys) {
      val = val?.[key];
    }
    return val ?? searchValue(data, value);
  }

  if (value.startsWith("$")) {
    let selector = value.slice(1);
    [selector, value] = parseFallback(selector);

    return queryCss(selector, data.page?.document) ?? searchValue(data, value);
  }

  return value;
}

function parseFallback(key: string): [string, string] {
  const fallback = key.indexOf("||");

  if (fallback !== -1) {
    return [
      key.slice(0, fallback).trim(),
      key.slice(fallback + 2).trim(),
    ];
  }

  return [
    key,
    "",
  ];
}

// https://regexr.com/7qnot
const checkForAttrPattern = /^(.+)\s+(?:attr\(([\w\-]+)\))$/;

function queryCss(query: string, document?: Document) {
  const checkResult = query.match(checkForAttrPattern);

  const hasAttr = checkResult?.[0];
  if (hasAttr) {
    const [_, query, name] = checkResult;
    return document?.querySelector(query)?.getAttribute(name);
  }

  return document?.querySelector(query)?.innerHTML;
}
