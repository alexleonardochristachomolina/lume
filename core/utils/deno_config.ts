import { parse } from "../../deps/jsonc.ts";
import { isUrl } from "../utils/path.ts";

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig extends Partial<ImportMap> {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx" | "precompile";
    jsxImportSource?: string;
    types?: string[];
  };
  [key: string]: unknown;
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      let importMap: ImportMap | undefined;

      if (config.importMap) {
        importMap = isUrl(config.importMap)
          ? await (await fetch(config.importMap)).json()
          : await JSON.parse(await Deno.readTextFile(config.importMap));
      } else if (config.imports) {
        importMap = {
          imports: config.imports,
          scopes: config.scopes,
        };
      }
      return { file, config, importMap };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}
