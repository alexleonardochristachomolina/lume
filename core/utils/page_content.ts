import type { Page } from "../file.ts";

export function insertContent(
  page: Page,
  content: string,
  placeholder?: string,
) {
  const pageContent = page.content;

  if (pageContent && typeof pageContent === "string") {
    if (placeholder && pageContent.includes(placeholder)) {
      page.content = pageContent.replace(
        placeholder,
        content,
      );
    } else {
      page.content += `\n${content}`;
    }
  } else {
    page.content = content;
  }
}
