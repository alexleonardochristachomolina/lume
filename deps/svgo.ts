export { optimize } from "npm:svgo@3.3.2";
export type { Config } from "npm:svgo@3.3.2";
