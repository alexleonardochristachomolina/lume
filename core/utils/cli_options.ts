import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { dev: "d", serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  const serveMode = cli.serve || cli._[0] === "cms";

  if (cli.port) {
    (options.server ||= {}).port = parseInt(cli.port);
  } else if (serveMode) {
    (options.server ||= {}).port = 3000;
  }

  let location: URL;

  if (cli.location) {
    location = new URL(cli.location);
  } else if (options.location && !serveMode) {
    location = options.location as URL;
  } else {
    location = new URL("http://localhost");
  }

  let port: number;

  if (cli.port) {
    port = parseInt(cli.port);
  } else if (location.port) {
    port = parseInt(location.port);
  } else if (options.server?.port) {
    port = options.server.port;
  } else {
    port = cli.serve ? 3000 : location.protocol === "https:" ? 443 : 80;
  }

  (options.server ||= {}).port = port;
  location.port = port.toString();
  options.location = location;

  if (cli.open) {
    (options.server ||= {}).open = cli.open;
  }

  return options;
}
