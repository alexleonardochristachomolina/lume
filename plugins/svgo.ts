import { optimize } from "../deps/svgo.ts";
import { merge } from "../core/utils/object.ts";

import type Site from "../core/site.ts";
import type { Page } from "../core/file.ts";
import type { Config } from "../deps/svgo.ts";

export interface Options {
  /** The list of extensions this plugin applies to */
  extensions?: string[];

  /** Options passed to SVGO. See https://github.com/svg/svgo#configuration */
  options?: Config;
}

// Default options
export const defaults: Options = {
  extensions: [".svg"],
};

/**
 * A plugin to load all SVG files and minify them using SVGO
 * @see https://lume.land/plugins/svgo/
 */
export function svgo(userOptions?: Options) {
  const options = merge(defaults, userOptions);

  return (site: Site) => {
    site.loadAssets(options.extensions);
    site.process(options.extensions, (pages) => pages.forEach(svg));

    function svg(page: Page) {
      const path = site.src(page.outputPath);
      const result = optimize(page.content as string, {
        path,
        ...options.options,
      }) as { data: string };

      page.content = result.data;
    }
  };
}

export default svgo;
