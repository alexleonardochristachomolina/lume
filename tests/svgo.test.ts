import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import svgo from "../plugins/svgo.ts";

Deno.test("terser plugin", async (t) => {
  const site = getSite({
    src: "svgo",
  });

  site.use(svgo());

  await build(site);
  await assertSiteSnapshot(t, site);
});
