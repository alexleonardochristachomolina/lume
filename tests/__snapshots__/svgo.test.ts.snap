export const snapshot = {};

snapshot[`terser plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      assetLoader: [AsyncFunction: text],
      engines: undefined,
      ext: ".svg",
      pageType: "asset",
    },
  ],
  src: [
    "/",
    "/favicon.svg",
  ],
}
`;

snapshot[`terser plugin 2`] = `[]`;

snapshot[`terser plugin 3`] = `
[
  {
    content: '<svg xmlns="http://www.w3.org/2000/svg" width="126" height="126"><path fill="#C84F46" fill-rule="evenodd" d="M52.868 7.995C48.907 24.097 69.167 21.154 69.167 44.72c6.628-7.615 1.28-17.875 11.891-27.547 0 22.213 21.097 20.964 12.954 52.052q5.663-3.47 8.143-11.425 5.004 7.956-3.738 28.279c5.13-2.164 8.141-5.156 9.961-9.352.663 29.238-18.872 49.64-46.391 49.268-27.52-.372-48.409-24.353-43.722-54.192.697 3.104 3.253 6.75 6.918 8.27-6.494-29.34 5.74-30.44 10.823-46.1-.826 8.324 0 12.952 4.015 17.786.71-2.112 1.994-4.12 3.351-6.131l.39-.575.196-.287.585-.865c2.784-4.138 5.281-8.413 3.205-13.732-3.379-8.658-3.057-12.024 5.12-22.174m-9.451 7.608c-2.174 2.218-1.702 8.672.797 14.34s-.797 9.933-4.329 15.184c.617-6.988-.026-8.647-3.026-14.02s1.11-12.484 6.558-15.504m21.355-7.57c5.083 4.416 6.886 4.857 7.437 9.404.552 4.548-1.912 6.66-.927 12.598-10.382-8.527-2.235-11.564-6.51-22.003M53.36 0c3.652 4.33 5.822 3.197 8.205 8.73 2.382 5.535-.602 6.79 1.436 13.445C51.003 11.853 58.446 11.017 53.359 0"/></svg>',
    data: {
      basename: "favicon",
      content: '<?xml version="1.0" encoding="UTF-8"?>
<svg width="126px" height="126px" viewBox="0 0 126 126" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <title>favicon</title>
    <g id="favicon" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <path d="M52.8683497,7.99450751 C48.9069025,24.0969979 69.1669662,21.1536885 69.1669662,44.7198348 C75.794917,37.1051328 70.4463322,26.8453748 81.058456,17.1725913 C81.058456,39.3863344 102.155357,38.1365645 94.0121919,69.2249916 C97.7866664,66.9118477 100.501055,63.103682 102.155357,57.8004944 C105.491398,63.103682 104.245257,72.5297487 98.416934,86.0786945 C103.547554,83.9150474 106.558165,80.9232746 108.378132,76.7271327 C109.04106,105.964869 89.5062156,126.366871 61.9867103,125.995028 C34.467205,125.623185 13.5784967,101.642443 18.2649214,71.802957 C18.9624768,74.9072731 21.5182769,78.5533566 25.1832626,80.0729101 C18.6891247,50.7321037 30.922502,49.6337285 36.0056643,33.9730813 C35.1796585,42.297059 36.0056643,46.924952 40.0208236,51.7593505 C40.7309276,49.6474958 42.0145668,47.6388953 43.3724725,45.627895 L43.7620351,45.0531669 C43.8271552,44.9573376 43.892336,44.86148 43.9575234,44.7655828 L44.5429346,43.9011285 C47.3268507,39.7628977 49.8240593,35.4880393 47.7482658,30.1694988 C44.3690672,21.5114095 44.6913719,18.1446262 52.8683497,7.99450751 Z M43.4168797,15.6028981 C41.2434833,17.8205255 41.7153274,24.275166 44.213695,29.9427578 C46.7120625,35.6103495 43.4168797,39.8756006 39.8854597,45.126676 C40.5019929,38.1390602 39.8587244,36.4802473 36.8588412,31.1077941 C33.858958,25.7353408 37.9685715,18.6230687 43.4168797,15.6028981 Z M64.7721431,8.03208837 C69.8550649,12.4485247 71.6576876,12.8901943 72.209291,17.4374541 C72.7608943,21.9847139 70.2967086,24.09635 71.2822168,30.0346208 C60.9002329,21.508497 69.0471936,18.4710785 64.7721431,8.03208837 Z M53.3587168,1.6350624e-05 C57.0113944,4.33013045 59.1813933,3.19675313 61.5638477,8.73089813 C63.9463021,14.2650431 60.9615342,15.5202483 63,22.1753596 C51.0029851,11.8529541 58.4461322,11.0173257 53.3587168,1.6350624e-05 Z" id="Combined-Shape-Copy" fill="#C84F46"></path>
    </g>
</svg>',
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/favicon.svg",
    },
    src: {
      asset: true,
      ext: ".svg",
      path: "/favicon",
      remote: undefined,
    },
  },
]
`;
